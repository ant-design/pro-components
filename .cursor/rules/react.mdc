---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
