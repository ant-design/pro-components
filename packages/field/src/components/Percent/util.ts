/** 获取展示符号 */
export function getSymbolByRealValue(realValue: number) {
  if (realValue === 0) {
    return null;
  }
  if (realValue > 0) {
    return '+';
  }
  return '-';
}

/** 获取颜色 */
export function getColorByRealValue(realValue: number /** ,color: string */) {
  if (realValue === 0) {
    return '#595959';
  }
  return realValue > 0 ? '#ff4d4f' : '#52c41a';
}

/** 获取到最后展示的数字 */
export function getRealTextWithPrecision(
  realValue: number,
  precision: number = 2,
) {
  return precision >= 0 ? realValue?.toFixed(precision) : realValue;
}

/**
 * 转化为数字
 * @copy from https://github.com/toss/es-toolkit/blob/32a183828c244d675f46810935e45dfefec81a54/src/compat/util/toNumber.ts#L19
 */
export function toNumber(value: any): number {
  if (typeof value === 'symbol' || value instanceof Symbol) {
    return NaN;
  }

  return Number(value);
}
