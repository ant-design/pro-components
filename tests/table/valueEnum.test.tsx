import ProProvider from '@ant-design/pro-provider';
import ProTable from '@ant-design/pro-table';
import { cleanup, render, waitFor } from '@testing-library/react';
import { Input } from 'antd';
import { act, useContext } from 'react';
import { waitForWaitTime } from '../util';

const Demo = () => {
  const values = useContext(ProProvider);
  return (
    <ProProvider.Provider
      value={{
        ...values,
        valueTypeMap: {
          link: {
            render: (text) => <a>{text}</a>,
            renderFormItem: (text, props) => (
              <Input placeholder="ËØ∑ËæìÂÖ•ÈìæÊé•" {...props?.fieldProps} />
            ),
          },
        },
      }}
    >
      <ProTable<any, Record<string, any>, 'link' | 'tags'>
        columns={[
          {
            title: 'ÈìæÊé•',
            dataIndex: 'name',
            valueType: 'link',
          },
        ]}
        request={() => {
          return Promise.resolve({
            total: 200,
            data: [
              {
                key: 1,
                name: 'test',
              },
            ],
            success: true,
          });
        }}
        rowKey="key"
      />
    </ProProvider.Provider>
  );
};

afterEach(() => {
  cleanup();
});

describe('Table valueEnum', () => {
  it('üéè dynamic enum test', async () => {
    const html = render(
      <ProTable
        size="small"
        columns={[
          {
            title: 'Áä∂ÊÄÅ',
            dataIndex: 'status',
            valueType: 'select',
            valueEnum: {},
            fieldProps: {
              open: true,
            },
          },
        ]}
        request={async () => ({
          data: [
            {
              status: 2,
              key: '1',
            },
          ],
        })}
        rowKey="key"
      />,
    );
    await waitFor(() => {
      return html.findAllByText('2');
    });

    act(() => {
      html.rerender(
        <ProTable
          size="small"
          request={async () => ({
            data: [
              {
                status: 2,
                key: '1',
              },
            ],
          })}
          rowKey="key"
          columns={[
            {
              title: 'Áä∂ÊÄÅ',
              valueType: 'select',
              dataIndex: 'status',
              valueEnum: {
                0: { text: 'ÂÖ≥Èó≠', status: 'Default' },
                1: { text: 'ËøêË°å‰∏≠', status: 'Processing', disabled: true },
                2: { text: 'Â∑≤‰∏äÁ∫ø', status: 'Success' },
                3: { text: 'ÂºÇÂ∏∏', status: 'Error' },
              },
              fieldProps: {
                open: true,
              },
            },
          ]}
        />,
      );
    });

    await waitFor(() => {
      return html.findAllByText('Â∑≤‰∏äÁ∫ø');
    });

    act(() => {
      html.baseElement
        .querySelector<HTMLDivElement>('form.ant-form div.ant-select')
        ?.click();
    });
    act(() => {
      expect(
        html.baseElement.querySelector<HTMLDivElement>(
          'div.ant-select-dropdown',
        )?.textContent,
      ).toBe('01ÂÖ≥Èó≠ËøêË°å‰∏≠Â∑≤‰∏äÁ∫øÂºÇÂ∏∏');
    });

    console.log(html.baseElement.querySelector('table')?.innerHTML);

    expect(
      html.baseElement.querySelector<HTMLDivElement>('td.ant-table-cell')
        ?.textContent,
    ).toBe('Â∑≤‰∏äÁ∫ø');
  });

  it('üéè customization valueType', async () => {
    const html = render(<Demo />);
    await waitForWaitTime(1200);
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('üéè dynamic request', async () => {
    const request = vi.fn();
    render(
      <ProTable
        size="small"
        columns={[
          {
            title: 'Áä∂ÊÄÅ',
            dataIndex: 'status',
            valueType: 'select',
            valueEnum: {},
            fieldProps: {
              open: true,
            },
            request: async (_, config) => {
              request(config.record);
              return [];
            },
          },
        ]}
        rowKey="key"
        request={async () => {
          return {
            data: [
              {
                status: 2,
                key: '1',
              },
            ],
          };
        }}
      />,
    );

    await waitFor(() => {
      expect(request).toHaveBeenCalledTimes(1);
    });
  });
});
