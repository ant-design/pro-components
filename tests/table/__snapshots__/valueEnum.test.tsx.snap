// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table valueEnum > 🎏 customization valueType 1`] = `<DocumentFragment />`;
